File: Leetcode/Clone_Binary_Tree_With_Random_Pointer.cpp
/**
 * Definition for a Node.
 * struct Node {
 *     int val;
 *     Node *left;
 *     Node *right;
 *     Node *random;
 *     Node() : val(0), left(nullptr), right(nullptr), random(nullptr) {}
 *     Node(int x) : val(x), left(nullptr), right(nullptr), random(nullptr) {}
 *     Node(int x, Node *left, Node *right, Node *random) : val(x), left(left), right(right), random(random) {}
 * };
 */

class Solution {
public:
    NodeCopy* copyRandomBinaryTree(Node* root) {
      if (root == NULL)
          return NULL;
        
      map<Node*, NodeCopy*> mapper;
      queue<Node*> qu1;
      queue<NodeCopy*> qu2;
      NodeCopy* rootDup(new NodeCopy(root->val));
      mapper[root] = rootDup;
      qu1.push (root);
        qu2.push (rootDup);
        while (!qu1.empty())
        {
            Node* curr = qu1.front();
            qu1.pop();
            NodeCopy* currDup = qu2.front ();
            qu2.pop();
            if (curr->left != NULL)
            {
                NodeCopy* leftChild (new NodeCopy (curr->left->val));
                mapper[curr->left] = leftChild;
                currDup->left = leftChild;
                qu1.push (curr->left);
                qu2.push(leftChild);  
            }
            if (curr->right != NULL)
            {
                NodeCopy* rightChild (new NodeCopy (curr->right->val));
                mapper[curr->right] = rightChild;
                currDup->right = rightChild;
                qu1.push (curr->right);
                qu2.push(rightChild);  
            }
        }
        qu1.push (root);
        qu2.push (rootDup);
        while (!qu1.empty())
        {
            Node* curr = qu1.front();
            qu1.pop();
            Node* randomptr = curr->random;
            NodeCopy* currEq = mapper [curr];
            NodeCopy* randomEq = mapper [randomptr];
            currEq->random = randomEq;
            if (curr->left != NULL)
                qu1.push (curr->left);            
            if (curr->right != NULL)
                qu1.push (curr->right);                        
        }
        return rootDup;
    }
};

File: Leetcode/Clone_Binary_Tree_With_Random_Pointer.java
/**
 * Definition for Node.
 * public class Node {
 *     int val;
 *     Node left;
 *     Node right;
 *     Node random;
 *     Node() {}
 *     Node(int val) { this.val = val; }
 *     Node(int val, Node left, Node right, Node random) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *         this.random = random;
 *     }
 * }
 */

class Solution {
    
    public NodeCopy copyRandomBinaryTree(Node root) {
        if (root ==null)
            return null;
        NodeCopy r = new NodeCopy (root.val);
        HashMap<Node, NodeCopy> mapper = new HashMap<Node, NodeCopy>();
        mapper.put (root,r);
        Queue<Node> qu = new LinkedList<Node>();
        Queue<NodeCopy> dup = new LinkedList<NodeCopy>();
        qu.add (root);
        dup.add (r);
        
        while (!qu.isEmpty())
        {
            Node curr = qu.poll();
            NodeCopy temp = dup.poll();
            if (curr.left != null)
            {
                qu.add (curr.left);
                NodeCopy child = new NodeCopy (curr.left.val);
                temp.left = child;
                mapper.put (curr.left, child);
                dup.add (child);               
            }
            
            if (curr.right != null)
            {
                qu.add (curr.right);
                NodeCopy child = new NodeCopy (curr.right.val);
                temp.right = child;
                mapper.put (curr.right, child);
                dup.add (child);               
            }   
        }
        qu.add (root);
        dup.add(r);
        while (!qu.isEmpty())
        {
            Node curr = qu.poll();
            NodeCopy tt = dup.poll();
            if (curr.random ==null)
                tt.random = null;
            else
            {
                NodeCopy ran =  mapper.get(curr.random);
                tt.random = ran;
            }
            if (curr.left != null)
            {
                qu.add (curr.left);
                dup.add (tt.left);    
            }
            
            if (curr.right != null)
            {
                qu.add (curr.right);
                dup.add (tt.right);    
            }
                        
        }
        return r;
    }
}
